Next gen

joliveros opened the pull request.

joliveros committed:
Update twilio-csharp to 5.0.0-rca2

joliveros committed:
Update controllers to use updated twilio sdk

joliveros committed:
Update tests

joliveros committed:
Add default AccountSid and authToken values to configuration

joliveros committed:
Update readme

smendes commented on LeadAlerts.Web/Domain/MessageSender.cs:
```diff
@@ -1,28 +1,21 @@
-            return _client.SendMessage(PhoneNumbers.Twilio, PhoneNumbers.Agent, message);
+            return await MessageResource.CreateAsync(
+                new PhoneNumber(PhoneNumbers.Twilio),
+                from: new PhoneNumber(PhoneNumbers.Agent), body: messageStr);
```
For better readability put `body` argument on another line

smendes commented on LeadAlerts.Web/Domain/MessageSender.cs:
```diff
@@ -1,28 +1,21 @@
 
 namespace LeadAlerts.Web.Domain
 {
-    public interface IMessageSender
```
I think you should keep an interface for MessageSender, it will be easier to mock

smendes commented on LeadAlerts.Web/Domain/MessageSender.cs:
```diff
@@ -1,28 +1,21 @@
+        public async Task<MessageResource> SendAsync(string messageStr)
         {
-            return _client.SendMessage(PhoneNumbers.Twilio, PhoneNumbers.Agent, message);
+            return await MessageResource.CreateAsync(
```
You should pass the `TwilioRestClient` as argument of `CreateAsync` to avoid an issue with the state of the client

acamino replied:
For consistency with the other implementations I would pass the TwilioRestClient as an argument to `MessageResource.CreateAsync`, this is a more OO approach. `Twilio.Init` is ok, although it relies in some kind of global state. I'd love to hear your thoughts on this.

joliveros committed:
Delete unnecessary tests

acamino commented on LeadAlerts.Web.Tests/Controllers/HomeControllerTest.cs:
```diff
@@ -12,7 +12,7 @@ public void GivenAIndexAction_ThenRendersDefaultView()
 
             controller.WithCallTo(c => c.Index())
-                .ShouldRenderDefaultView();
+                .ShouldRenderView("Index");
```
Why did you use `ShouldRenderView(...)`? You're calling the action method `Index`, the default view for that one is "Index". `ShouldRenderDefaultView` has the necessary semantics to express the intention of the test. What do you think?

joliveros replied:
In my head it seemed to make more sense. But as you put it that way, `ShouldRenderDefaultView` is much better. Thanks for clarifying that!

acamino commented on LeadAlerts.Web/Domain/Configuration.cs:
```diff
@@ -6,12 +6,14 @@ public class Credentials
         {
-            get { return WebConfigurationManager.AppSettings["TwilioAccountSid"]; }
+            get { return WebConfigurationManager.AppSettings["TwilioAccountSid"] ??
+                    "ACXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"; }
```
Why did you add a fallback value? With this value the application will return a resource not found (404), so I can't see a benefit on this approach. Is there something I'm missing here?

joliveros replied:
Ah yes, this is a vestige from when I had attempted to test the twilio api calls directly instead of mocking the client wrapper. I should remove this...

acamino commented on LeadAlerts.Web/Domain/Configuration.cs:
```diff
@@ -6,12 +6,14 @@ public class Credentials
         {
-            get { return WebConfigurationManager.AppSettings["TwilioAuthToken"]; }
+            get { return WebConfigurationManager.AppSettings["TwilioAuthToken"] ?? 
+                    "aXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"; }
```
ditto

acamino commented on LeadAlerts.Web/Domain/MessageSender.cs:
```diff
@@ -1,28 +1,28 @@
     public interface IMessageSender
     {
-        Message Send(string message);
+        Task<MessageResource> SendAsync(string messageStr); 
```
As a good practice we don't use the type of the variable as a postfix token (that was common when you use Hungarian notation). This parameter should be `message`. Does it make sense to you?

joliveros replied:
Yes, of course. Nice catch!

acamino replied:
Would you mind changing it please?

joliveros committed:
Update MessageSender with better client init, minor code style changes

acamino commented on LeadAlerts.Web.Tests/Controllers/NotificationsControllerTest.cs:
```diff
@@ -18,14 +21,18 @@ public void GivenACreateAction_ThenItSendsAMessage()
 
             controller.WithCallTo(c => c.Create(new Lead()))
-                .ShouldRedirectTo<HomeController>(home => home.Index());
+                .ShouldRedirectToRoute("");
```
The previous version was more readable. Please rollback this change.

acamino commented on LeadAlerts.Web/Controllers/NotificationsController.cs:
```diff
@@ -2,27 +2,28 @@
             _messageSender = messageSender;
         }
-
+        
```
No need to add spaces here, it's supposed to be an empty line.

acamino reviewed (commented):
Just a few changes more, and we're ready.

acamino committed:
Refactor tests to make them more explicit

acamino committed:
Refactor `Configuration.cs` to use expression body

acamino committed:
User string instead of String

acamino committed:
Refactor: Rename MessageSender to NotificationService

Update tests to gain confidence. Usually it's better to write an assertion
against to a concrete value instead of "any" string.

acamino commented:
@joliveros looks like this one is ready to get merged.

acamino committed:
Remove `ace` folder

acamino committed:
Update README to include `Local.config.example` instructions

joliveros commented:
LGTM @acamino

smendes committed:
Update twilio-csharp to latest stable 5.0.0
